cvxpylayers/jax/cvxpylayer.py

```diff
@@ -0,0 +1,322 @@
+import numpy as np
+import time
+from functools import partial
+
+try:
+    import jax
+except ImportError:
+    raise ImportError("Unable to import jax. Please install from "
+                      "https://github.com/google/jax")
+from jax import core
+import jax.numpy as jnp
+
+
+def CvxpyLayer(problem, parameters, variables, gp=False):
+    """Construct a CvxpyLayer
+
+    Args:
+        problem: The CVXPY problem; must be DPP.
+        parameters: A list of CVXPY Parameters in the problem; the order
+                    of the Parameters determines the order in which parameter
+                    values must be supplied in the forward pass. Must include
+                    every parameter involved in problem.
+        variables: A list of CVXPY Variables in the problem; the order of the
+                   Variables determines the order of the optimal variable
+                   values returned from the forward pass.
+        gp: Whether to parse the problem using DGP (True or False).
+
+    Returns:
+        A callable that solves the problem.
+    """
+
+    if gp:
+        if not problem.is_dgp(dpp=True):
+            raise ValueError('Problem must be DPP.')
+    else:
+        if not problem.is_dcp(dpp=True):
+            raise ValueError('Problem must be DPP.')
+
+    if not set(problem.parameters()) == set(parameters):
+        raise ValueError("The layer's parameters must exactly match "
+                         "problem.parameters")
+    if not set(variables).issubset(set(problem.variables())):
+        raise ValueError("Argument variables must be a subset of "
+                         "problem.variables")
+    if not isinstance(parameters, list) and \
+            not isinstance(parameters, tuple):
+        raise ValueError("The layer's parameters must be provided as "
+                         "a list or tuple")
+    if not isinstance(variables, list) and \
+            not isinstance(variables, tuple):
+        raise ValueError("The layer's variables must be provided as "
+                         "a list or tuple")
+
+    var_dict = {v.id for v in variables}
+
+    # Construct compiler
+    param_order = parameters
+    if gp:
+        for param in parameters:
+            if param.value is None:
+                raise ValueError("An initial value for each parameter is "
+                                 "required when gp=True.")
+        data, solving_chain, _ = problem.get_problem_data(
+            solver=cp.SCS, gp=True)
+        compiler = data[cp.settings.PARAM_PROB]
+        dgp2dcp = solving_chain.get(cp.reductions.Dgp2Dcp)
+        param_ids = [p.id for p in compiler.parameters]
+        old_params_to_new_params = (
+            dgp2dcp.canon_methods._parameters
+        )
+    else:
+        data, _, _ = problem.get_problem_data(solver=cp.SCS)
+        compiler = data[cp.settings.PARAM_PROB]
+        param_ids = [p.id for p in param_order]
+        dgp2dcp = None
+    cone_dims = dims_to_solver_dict(data["dims"])
```

Comment:
At some point we should abstract this into a common function, used by all three layers. Maintaining three separate versions of the same code won't be fun.

No pressure to do this now (though you're more than welcome to do so, if you want to).